### ISSUE TITLE
FieldPanel.get_comparison_class() doesn't seem to act normally

### ISSUE THREAD

**Issue by Zogsha** [wagtail/wagtail#5130]
My previous messages were not clear at all. Sorry, here is a detailed issue

### Issue Summary
When you compare revisions with `CharField`, `TextField `or `RichTextField`, the output doesn't seem right.
`FieldPanel.get_comparison_class()` doesn't catch `CharField `and `TextField`, `diff_text`() doesn't work the right way.

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. Add a TextField and a RichTextField to the HomePage
3. Edit it and compare revisions

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes

### Technical details

* Python version: 3.7.2
* Django version: 2.1.7
* Wagtail version: 2.4

To solve the problem edited edit_handler.py and compare.py in wagtail/admin/
First I added a new condition at the bottom of `FieldPanel.get_comparison_class()` in edit_handler.py to catch `TextField `and `CharField `and `return compare.TextFieldComparison` if `field `is an instance of one these.
Then in compare.py, I inverted val_a and val_b in `RichTextFieldComparison.htmldiff()` and `TextFieldComparison.htmldiff()`.
Finally I inverted `'addition'` and `'deletion'` in `diff_text()`

All together:
```python
# wagtail/admin/edit_handler.py

# New Imports
from django.db.models.field import TextField, CharField

class FieldPanel(EditHandler):
    ...

    def get_comparison_class(self):
        # Hide fields with hidden widget
        widget_override = self.widget_overrides().get(self.field_name, None)
        if widget_override and widget_override.is_hidden:
            return

        try:
            field = self.db_field

            if field.choices:
                return compare.ChoiceFieldComparison

            if field.is_relation:
                if isinstance(field, TaggableManager):
                    return compare.TagsFieldComparison
                elif field.many_to_many:
                    return compare.M2MFieldComparison

                return compare.ForeignObjectComparison

            if isinstance(field, RichTextField):
                return compare.RichTextFieldComparison

            # New check - if placed before RichTextField it catches the RichTextField too
            if isinstance(field, TextField) or isinstance(field, CharField):
                return compare.TextFieldComparison
    
        except FieldDoesNotExist:
            pass

        return compare.FieldComparison

    ...

# wagtail/admin/compare.py

# invert val_b and val_a
class TextFieldComparison(FieldComparison):
    def htmldiff(self):
        return diff_text(self.val_b, self.val_a).to_html()

# invert val_b and val_a
class RichTextFieldComparison(TextFieldComparison):
    def htmldiff(self):
        return diff_text(
            BeautifulSoup(force_text(self.val_b), 'html5lib').getText(),
            BeautifulSoup(force_text(self.val_a), 'html5lib').getText()
        ).to_html()


# invert addition and deletion
def diff_text(a, b):

    def tokenise(text):
        ...

    a_tok = tokenise(a)
    b_tok = tokenise(b)
    sm = difflib.SequenceMatcher(lambda t: len(t) <= 4, a_tok, b_tok)

    changes = []

    for op, i1, i2, j1, j2 in sm.get_opcodes():
        if op == 'replace':
            for token in a_tok[i1:i2]:
                changes.append(('addition', token))
            for token in b_tok[j1:j2]:
                changes.append(('deletion', token))
        elif op == 'delete':
            for token in a_tok[i1:i2]:
                changes.append(('addition', token))
        elif op == 'insert':
            for token in b_tok[j1:j2]:
                changes.append(('deletion', token))
        elif op == 'equal':
            for token in a_tok[i1:i2]:
                changes.append(('equal', token))
    
    ...
```

the different outputs for the different steps are as follow:
![comparison_wagtail](https://user-images.githubusercontent.com/17270547/54094670-ebb85580-43a2-11e9-8e2f-1d1da90a91c6.png)

I hope it is clear enough this time.

**Comment by kaedroho** [comment]
Just been trying these changes out locally, definitely gives a big improvement!

I've submitted these changes as a PR at https://github.com/wagtail/wagtail/pull/5132

Thanks!

**Comment by Zogsha** [comment]
Thank you for your interest.
I tested this morning with a new fresh wagtail install, without modifying `diff_text()` function, only adding the `TextField `and `CharField `check in **wagtail/admin/edit_handlers.py** `FieldPanel.get_comparison_class()` and the result corresponded to the tests assertions in **wagtail/admin/tests/test_compare.py**.
![text_diffing](https://user-images.githubusercontent.com/17270547/54188539-3247a600-44b0-11e9-8abf-3d09f4a4abdf.png)
The `FieldPanel.get_comparison_class()` modification seems necessary. But now I am wondering if I really had a fresh install for my first message. I probably mixed up my virtual envs and with an already modified `diff_text()`.
I think there is no need to invert values in `RichTextFieldComparison.htmldiff()` and `TextFieldComparison.htmldiff()` and `diff_text()` function, as i mentioned in my first message.
I apologize.

**Comment by kaedroho** [comment]
Thanks for clarifying @Zogsha! I'll update the MR.

**Comment by chosak** [comment]
Fix in #1532 has been merged in 40dbb94. Thanks again @Zogsha!
